File: app/src/main/java/com/example/pie/android/MainActivity.java
package com.example.pie.android;

import android.os.Bundle;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.Toast;

import com.example.pie.android.adapter.TodoAdapter;
import com.example.pie.android.model.TodoItem;
import com.example.pie.android.rest.resource.TodoResource;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Consumer;

public class MainActivity extends AppCompatActivity implements SwipeRefreshLayout.OnRefreshListener {

    private ArrayAdapter<TodoItem> adapter;
    private SwipeRefreshLayout refresh;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        refresh = (SwipeRefreshLayout) findViewById(R.id.refresh);
        refresh.setOnRefreshListener(this);
        adapter = new TodoAdapter(this, R.layout.todo_item, new ArrayList<TodoItem>() );

        final ListView list = (ListView) findViewById(R.id.list);
        list.setAdapter(adapter);
        retrieveItems();
    }

    public void retrieveItems() {
        new TodoResource().getItems().subscribe(new Consumer<List<TodoItem>>() {
            @Override
            public void accept(@NonNull List<TodoItem> todoItems) throws Exception {
                // This should probably be made more efficient.
                adapter.clear();
                adapter.addAll(todoItems);
                refresh.setRefreshing(false);
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(@NonNull Throwable throwable) throws Exception {
                refresh.setRefreshing(false);
                Toast.makeText(MainActivity.this, "Failed to contact Server.", Toast.LENGTH_SHORT).show();
                throwable.printStackTrace();
            }
        });
    }

    public void submitItem(View v){
        final EditText input = (EditText) findViewById(R.id.newTask);
        String userInput = input.getText().toString().trim();
        if(userInput.length() > 0){
            TodoItem newItem = new TodoItem();
            newItem.setTask(userInput);
            List<TodoItem> newItems = new ArrayList<>();
            newItems.add(newItem);

            new TodoResource().addItems(newItems).subscribe(new Consumer<List<TodoItem>>() {
                @Override
                public void accept(@NonNull List<TodoItem> todoItems) throws Exception {
                    adapter.addAll(todoItems);
                    input.setText("");
                }
            }, new Consumer<Throwable>() {
                @Override
                public void accept(@NonNull Throwable throwable) throws Exception {
                    Toast.makeText(MainActivity.this, "Failed to contact Server.", Toast.LENGTH_SHORT).show();
                    throwable.printStackTrace();
                }
            });
        }
    }

    public void clearCompletedTasks(View v) {
        final EditText input = (EditText) findViewById(R.id.newTask);
        new TodoResource().clearDone().subscribe(new Consumer<List<TodoItem>>() {
            @Override
            public void accept(@NonNull List<TodoItem> todoItems) throws Exception {
                adapter.clear();
                adapter.addAll(todoItems);
                input.setText("");
            }
        }, new Consumer<Throwable>() {
            @Override
            public void accept(@NonNull Throwable throwable) throws Exception {
                throwable.printStackTrace();
            }
        });
    }

    @Override
    public void onRefresh() {
        retrieveItems();
    }

    @Override
    public void onRestart() {
        super.onRestart();
        retrieveItems();
    }
}

File: app/src/main/java/com/example/pie/android/MainActivity.kt
package com.example.pie.android

import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout

import com.example.pie.android.adapter.TodoAdapter
import com.example.pie.android.databinding.ActivityMainBinding
import com.example.pie.android.model.TodoItem
import com.example.pie.android.rest.TodoResource
import dagger.android.support.DaggerAppCompatActivity

import java.util.ArrayList

import javax.inject.Inject

class MainActivity : DaggerAppCompatActivity(), SwipeRefreshLayout.OnRefreshListener {

    private lateinit var adapter: ArrayAdapter<TodoItem>
    private lateinit var binding: ActivityMainBinding
    private val items = ArrayList<TodoItem>()
    @Inject lateinit var resource: TodoResource

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.refresh.setOnRefreshListener(this)
        adapter = TodoAdapter(resource, this, R.layout.todo_item, items)

        binding.list.adapter = adapter
        if (savedInstanceState != null) {
            adapter.addAll(savedInstanceState.getParcelableArrayList(ITEMS_PARAM) ?: emptyList())
        } else {
            retrieveItems()
        }
    }

    private fun retrieveItems() {
        resource.items.subscribe({ todoItems ->
            // This should probably be made more efficient.
            adapter.clear()
            adapter.addAll(todoItems)
            binding.refresh.isRefreshing = false
        }) { throwable ->
            binding.refresh.isRefreshing = false
            Toast.makeText(this, "Failed to contact Server.", Toast.LENGTH_SHORT).show()
            throwable.printStackTrace()
        }
    }

    fun submitItem(v: View) {
        val userInput = binding.newTask.text.toString().trim { it <= ' ' }
        if (userInput.isNotEmpty()) {
            val newItem = TodoItem()
            newItem.task = userInput
            val newItems = ArrayList<TodoItem>()
            newItems.add(newItem)

            resource.addItems(newItems).subscribe({ todoItems ->
                adapter.addAll(todoItems)
                binding.newTask.setText("")
            }) { throwable ->
                Toast.makeText(this, "Failed to contact Server.", Toast.LENGTH_SHORT).show()
                throwable.printStackTrace()
            }
        }
    }

    fun clearCompletedTasks(v: View) {
        resource.clearDone().subscribe({ todoItems ->
            adapter.clear()
            adapter.addAll(todoItems)
            binding.newTask.setText("")
        }) { throwable -> throwable.printStackTrace() }
    }

    override fun onRefresh() {
        retrieveItems()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelableArrayList(ITEMS_PARAM, items)
    }

    companion object {
        private const val ITEMS_PARAM = "items"
    }
}
